agent/consul/multilimiter/multilimiter.go

```diff
@@ -0,0 +1,188 @@
+type MultiLimiter struct {
+	limiters *atomic.Pointer[radix.Tree]
+	config   *atomic.Pointer[Config]
+}
+
+type keyType = []byte
+
+// LimitedEntity is an interface used by MultiLimiter.Allow to determine
+// which rate limiter to use to allow the request
+type LimitedEntity interface {
+	Key() keyType
+}
+
+// Limiter define a limiter to be part of the MultiLimiter structure
+type Limiter struct {
+	limiter    *rate.Limiter
+	lastAccess atomic.Int64
+	config     *atomic.Pointer[LimiterConfig]
+}
+
+// LimiterConfig is a Limiter configuration
+type LimiterConfig struct {
+	Rate  rate.Limit
+	Burst int
+}
+
+// Config is a MultiLimiter configuration
+type Config struct {
+	LimiterConfig
+	CleanupCheckLimit    time.Duration
+	CleanupCheckInterval time.Duration
+}
+
+func (c *Config) Equal(lc *LimiterConfig) bool {
+	return c.Burst == lc.Burst && c.Rate == lc.Rate
+}
+
+// UpdateConfig will update the MultiLimiter Config
+// which will cascade to all the Limiter(s) LimiterConfig
+func (m *MultiLimiter) UpdateConfig(c LimiterConfig, prefix []byte) {
+	newLimiters, _, _ := m.limiters.Load().Insert(prefix, c)
+	m.limiters.Store(newLimiters)
+}
+
+// NewMultiLimiter create a new MultiLimiter
+func NewMultiLimiter(c Config) *MultiLimiter {
+	limiters := atomic.Pointer[radix.Tree]{}
+	config := atomic.Pointer[Config]{}
+	config.Store(&c)
+	limiters.Store(radix.New())
+	if c.CleanupCheckLimit == 0 {
+		c.CleanupCheckLimit = 30 * time.Millisecond
+	}
+	if c.CleanupCheckInterval == 0 {
+		c.CleanupCheckLimit = 1 * time.Second
+	}
+	m := &MultiLimiter{limiters: &limiters, config: &config}
+	return m
+}
+
+// Run the cleanup routine to remove old entries of Limiters based on CleanupCheckLimit and CleanupCheckInterval.
+func (m *MultiLimiter) Run(ctx context.Context) {
+	go func() {
+		for {
+			m.cleanupLimitedOnce(ctx)
+		}
+	}()
+}
+
+// todo: split without converting to a string
+func splitKey(key []byte) ([]byte, []byte) {
+
+	s := strings.Split(string(key), string(separator))
+	if len(s) >= 2 {
+		return []byte(s[0]), []byte(s[1])
+	}
+	return nil, nil
+}
+
+// Allow should be called by a request processor to check if the current request is Limited
+// The request processor should provide a LimitedEntity that implement the right Key()
+func (m *MultiLimiter) Allow(e LimitedEntity) bool {
+	prefix, _ := splitKey(e.Key())
+	limiters := m.limiters.Load()
+	l, ok := limiters.Get(e.Key())
+	now := time.Now().Unix()
+	if ok {
+		limiter := l.(*Limiter)
+		limiter.lastAccess.Store(now)
+		return limiter.limiter.Allow()
+	}
+	c, okP := limiters.Get(prefix)
+	var config LimiterConfig
+	if okP {
+		config = c.(LimiterConfig)
+	} else {
+		config = m.config.Load().LimiterConfig
+	}
+	limiter := &Limiter{limiter: rate.NewLimiter(config.Rate, config.Burst), config: &atomic.Pointer[LimiterConfig]{}}
+	limiter.config.Store(&config)
+	limiter.lastAccess.Store(now)
+	tree, _, _ := limiters.Insert(e.Key(), limiter)
```

Comment:
As an aside/not something we need to change now: I think the structure of the tree will need to change slightly in the future to accommodate setting config directly on a leaf-node (e.g. for a specific identity/tenant).

I guess we'd store a leaf struct containing the config and limiter? We can figure that out when we get to it, though 😄

Reply:
yes that was one of the option I thought about, I think it make sense to make the tree store:
```go
type not_sure_about_the_name struct {
    config *LimiterConfig
    limiter *Limiter
} 
```
and in the cleanup routine (I think at this point we need to renamed to reconcile), we check if any is nil, instead of the type check.

Reply:
This is not needed anymore as config and limiters are now separate trees which make that case natively supported